homeassistant/components/sensor/foobot.py

```diff
@@ -0,0 +1,151 @@
+    vol.Required(CONF_TOKEN): cv.string,
+    vol.Required(CONF_USERNAME): cv.string,
+})
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up the devices associated with the account."""
+    from foobot_async import FoobotClient
+
+    token = config.get(CONF_TOKEN)
+    username = config.get(CONF_USERNAME)
+
+    client = FoobotClient(token, username,
+                          async_get_clientsession(hass),
+                          timeout=TIMEOUT)
+    data = {}
+    dev = []
+    try:
+        devices = yield from client.get_devices()
+        _LOGGER.debug("The following devices were found: %s", devices)
+        for device in devices:
+            for sensor_type in SENSOR_TYPES.keys():
+                if sensor_type == 'time':
+                    continue
+                foobot_sensor = FoobotSensor(client, data, device,
+                                             sensor_type)
+                dev.append(foobot_sensor)
+    except (aiohttp.client_exceptions.ClientConnectorError,
+            asyncio.TimeoutError, FoobotClient.TooManyRequests,
+            FoobotClient.InternalError):
+        _LOGGER.exception('Failed to connect to foobot servers.')
+        raise PlatformNotReady
+    except (FoobotClient.ClientError):
+        _LOGGER.exception('Failed to fetch data from foobot servers.')
+        raise HomeAssistantError
+    async_add_devices(dev, True)
+
+
+class FoobotSensor(Entity):
+    """Implementation of a Foobot sensor."""
+
+    __cache = {}
+
+    def __init__(self, client, data, device, sensor_type):
+        """Initialize the sensor."""
+        self._client = client
+        self._uuid = device['uuid']
+        self._name = 'Foobot {} {}'.format(device['name'],
+                                           SENSOR_TYPES[sensor_type][0])
+        self.type = sensor_type
+        self._unit_of_measurement = SENSOR_TYPES[sensor_type][1]
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend."""
+        return SENSOR_TYPES[self.type][2]
+
+    @property
+    def state(self):
+        """Return the state of the device."""
+        try:
+            data = self.__cache[self._uuid][self.type]
+        except(KeyError, TypeError):
+            data = None
+        return data
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement of this entity."""
+        return self._unit_of_measurement
+
+    @asyncio.coroutine
+    def async_update(self):
+        """Get the latest data."""
+        if self._uuid in self.__cache:
+            last_request = self.__cache[self._uuid]['last_request']
+            _LOGGER.debug('last_update: %s', last_request)
+            _LOGGER.debug('now: %s', datetime.utcnow())
+            if last_request + SCAN_INTERVAL > datetime.utcnow():
```

Comment:
Home Assistant understands your SCAN_INTERVAL variable and will call your code every SCAN_INTERVAL. Manual checking is not needed

Reply:
The issue I ran into and that necessitated this manual check is that each sensor is its own entity, 6 in total.

SCAN_INTERVAL is indeed respected but for each sensor independently, it'll try to make 6 API calls every SCAN_INTERVAL. This code is for checking if another sensor didn't already make an API call and get the data if it's available.

I debated having only one entity and putting the different sensor values into attributes but I choose the current 6 independent sensors for ease of use for the end user.

There is probably a more elegant way of doing this. I'm open to suggestions.